lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,548 @@
+  assert(updatedInputs.size() > fieldIndex.getValue() &&
+         "flattened array small");
+
+  op.replaceAllUsesWith(updatedInputs[fieldIndex.getValue()].first);
+  op.erase();
+  return true;
+}
+
+bool DoLowering::lowerConnectOp(ConnectOp op) {
+  Value dest = op.dest();
+  Value src = op.src();
+
+  // Attempt to get the bundle types, potentially unwrapping an outer flip
+  // type that wraps the whole bundle.
+  FIRRTLType destType = getCanonicalAggregateType(dest.getType());
+  FIRRTLType srcType = getCanonicalAggregateType(src.getType());
+
+  // If we aren't connecting two bundles, there is nothing to do.
+  if (!destType || !srcType)
+    return false;
+
+  LoweredValues destValues;
+  auto l1 = getBundleLowering(dest, destValues);
+  assert(l1 && "Not yet lowered");
+
+  // allVector<std::pair<Value, bool>, 8> srcValues;
+  LoweredValues srcValues;
+  auto l2 = getBundleLowering(src, srcValues);
+  assert(l2 && "not yet lowered");
+
+  for (auto tuple : llvm::zip_first(destValues, srcValues)) {
+
+    auto newDest = std::get<0>(tuple).first;
+    auto newDestFlipped = std::get<0>(tuple).second;
+    auto newSrc = std::get<1>(tuple).first;
+
+    // Flow checks guarantee that the connection is valid.  Therfore,
+    // no flow checks are needed and just the type of the LHS
+    // determines whether or not this is a reverse connection.
+    if (newDestFlipped)
+      std::swap(newSrc, newDest);
+
+    builder.create<ConnectOp>(newDest, newSrc);
+  }
+  op.erase();
+  return true;
+}
+
+void DoLowering::runLowering() {
+  // Iterate till the worklist is empty.
+  while (!worklist.empty()) {
+    auto item = worklist.front();
+    worklist.erase(worklist.begin());
+    // Ignore if the operation is already lowered.
+    if (isLowered(item))
+      continue;
+    // Step 1: Lower the operation which generates an aggregate type result.
+    // ==
+    // Special handling of BlockArguments. 
+    if (auto arg = item.dyn_cast<BlockArgument>()) {
+      builder.setInsertionPoint(module);
+      builder.setLoc(module.getLoc());
+      lowerArg(arg);
+    }
+    // Step 2: After lowering, replace all uses of the item with the lowered values. 
+    //         ( For certain ops like ConnectOp, we cannot replace an operand with the lowered values, unless all the source operands of the operation are lowered.
+    // ==
+    bool allUsersRemoved = true;
+    for (auto user : item.getUsers()) {
+      auto &u = *user;
+      builder.setInsertionPoint(&u);
+      builder.setLoc(u.getLoc());
+      bool opRemoved =
+          TypeSwitch<Operation *, bool>(&u)
+              .Case<ConnectOp>([&](auto op) {
+                bool ready = true;
+                for (auto operand : op.getOperands()) {
+                  if (!isLowered(operand)) {
+                    ready = false;
+                    worklist.push_back(operand);
+                  }
+                }
+                if (ready)
+                  return lowerConnectOp(op);
```

Comment:
lowerConnect can check if it's ready, they you would have a uniform dispatch which could be done with a visitor